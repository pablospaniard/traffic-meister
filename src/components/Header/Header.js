import React, { useState, useContext, useEffect } from 'react'
import isEmpty from 'lodash/isEmpty'

import { FlexContainer, FlexItem, Select } from '../UI'
import AppContext from '../../helpers/context'
import { PROPS } from '../../helpers/constants'

let arrayOfTypes = []
let arrayOfBrands = []
let arrayOfColors = []

const Header = () => {
  const { data, filters, setFilters, filteredData } = useContext(AppContext)
  // this could be done with useReducer, but i leave it as is for simplicity
  const [state, setState] = useState({})

  useEffect(() => {
    if (!isEmpty(data)) {
      const source = isEmpty(filteredData) ? data : filteredData
      arrayOfTypes = source.map(i => i.type)
      const uniqueTypes = Array.from(new Set(arrayOfTypes))
      arrayOfBrands = source.map(i => i.brand)
      const uniqueBrands = Array.from(new Set(arrayOfBrands))
      arrayOfColors = source.reduce(
        (prev, curr) => [...prev, ...curr.colors],
        []
      )
      const uniqueColors = Array.from(new Set(arrayOfColors))
      setState({
        ...state,
        types: uniqueTypes,
        brands: uniqueBrands,
        colors: uniqueColors
      })
    }
  }, [data, filteredData])

  const handleChange = (item, prop) => {
    switch (prop) {
      case PROPS.TYPES:
        setFilters({ ...filters, type: item })
        break
      case PROPS.BRANDS:
        setFilters({ ...filters, brand: item })
        break
      case PROPS.COLORS:
        setFilters({ ...filters, color: item })
        break

      default:
        break
    }
  }

  return (
    <>
      <h1>The Traffic Meister</h1>
      <form>
        <FlexContainer>
          <FlexItem flex="1" margin="0 20px">
            <Select
              value={filters.type}
              items={state.types}
              handleChange={item => handleChange(item, 'type')}
              disabled={isEmpty(state.types)}
              loading={isEmpty(data)}
            />
          </FlexItem>
          <FlexItem flex="1" margin="0 20px">
            <Select
              value={filters.brand}
              items={state.brands}
              handleChange={item => handleChange(item, 'brand')}
              disabled={isEmpty(state.brands)}
              loading={isEmpty(data)}
            />
          </FlexItem>
          <FlexItem flex="1" margin="0 20px">
            <Select
              value={filters.color}
              items={state.colors}
              handleChange={item => handleChange(item, 'color')}
              disabled={isEmpty(state.colors)}
              loading={isEmpty(data)}
            />
          </FlexItem>
        </FlexContainer>
      </form>
    </>
  )
}

export default Header
