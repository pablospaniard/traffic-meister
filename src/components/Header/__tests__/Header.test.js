import React from 'react'
import { render } from '@testing-library/react'

import Header from '../Header'
import AppContext from '../../../helpers/context'

const filters = {
  type: { label: 'car', value: 'Car' },
  brand: { label: 'yokohama', value: 'Yokohama' },
  color: { label: 'silver', value: 'silver' }
}
const data = [
  {
    id: 1,
    type: 'car',
    brand: 'Bugatti Veyron',
    colors: ['red', 'black'],
    img:
      'https://upload.wikimedia.org/wikipedia/commons/thumb/c/c9/Bugatti_Veyron_16.4_%E2%80%93_Frontansicht_%281%29%2C_5._April_2012%2C_D%C3%BCsseldorf.jpg/520px-Bugatti_Veyron_16.4_%E2%80%93_Frontansicht_%281%29%2C_5._April_2012%2C_D%C3%BCsseldorf.jpg'
  },
  {
    id: 2,
    type: 'airplane',
    brand: 'Boeing 787 Dreamliner',
    colors: ['red', 'white', 'black', 'green'],
    img:
      'https://upload.wikimedia.org/wikipedia/commons/thumb/1/15/All_Nippon_Airways_Boeing_787-8_Dreamliner_JA801A_OKJ_in_flight.jpg/600px-All_Nippon_Airways_Boeing_787-8_Dreamliner_JA801A_OKJ_in_flight.jpg'
  }
]

describe('Header component should works properly', () => {
  it('should match Snapshot', () => {
    const tree = (
      <AppContext.Provider value={{ data, filters }}>
        <Header />
      </AppContext.Provider>
    )
    const { container } = render(tree)
    expect(container.firstChild).toMatchSnapshot('Header_1')
  })
})
