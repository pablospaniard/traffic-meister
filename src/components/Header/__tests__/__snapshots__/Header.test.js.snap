// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component should works properly should match Snapshot: Header_1 1`] = `
<h1
  style="text-align: center;"
>
  The Traffic Meister
</h1>
`;
